### DISCUSSION TITLE
Detected multiple Jotai instances. It may cause unexpected behavior with the default store.

### DISCUSSION THREAD

**Discussion by fabio-nettis** [pmndrs/jotai#2072] [Bug report] [answered]
I suddenly am encountering this error after upgrading to Next.js v13.4.13 from the previous version. My root layout.tsx file includes a provider for Jotai, thats the only place in the application the provider can be found. Here is the code of the provider component:

```tsx
"use client";

import { Provider } from "jotai";
import store from "global/store";

interface Props extends React.PropsWithChildren {}

export default function JotaiProvider({ children }: Props): JSX.Element {
  return <Provider store={store}>{children}</Provider>;
}

export type { Props as JotaiProviderProps };
```

Here I am simply passing the store created in `"global/store"` and wrapping the rest of the children with it. Here is the code that creates the store.

```tsx
import { getDefaultStore } from "jotai";

const store = getDefaultStore();

export default store;
```

Am I doing something wrong here? This error may be related to the fix #2066, but again that is just a wild guess.

**Comment by dai-shi** [comment, marked as answer]
That's exactly the case the warning is trying to detect.
Using getDefaultStore() assumes single Jotai instance.

If you are using Nextjs, it's recommended to avoid using the default store.
```diff
-   return <Provider store={store}>{children}</Provider>;
+   return <Provider>{children}</Provider>;
```

**Reply by fabio-nettis** [reply]
That did the job, my bad, thank you.

**Reply by anshul-kai** [reply]
Removing the `store` prop from the `Provider` throws the following error. Adding it removes it. Is the above fix still valid for Next.js projects?

`Detected multiple Jotai instances. It may cause unexpected behavior with the default store. https://github.com/pmndrs/jotai/discussions/2044`

Regardless, it looks like there is more than one store as updates aren't always propagating to other components as one would expect.

**Reply by IsmailKemmoune** [reply]
@dai-shi I have a question please, doesn't wrapping the root layout in case of next js app router and marking it as a client component, makes my whole app a client component? since all the sub components gonna be client as well?

**Reply by dai-shi** [reply]
Yeah, that's why you want to use `Provider` in a server component. Or, you can create your own client component that receives a server component as a children prop, which avoids making the whole app a client one.

**Comment by dai-shi** [comment]
@anshul-kai 

> Is the above fix still valid for Next.js projects?

It depends on your requirement, but in general, SSR framework like Next.js shares memory across multiple sessions and you want to make sure to use different store for each session. Removing `store` prop will create a new store for each session.

**Reply by anshul-kai** [reply]
Thank you @dai-shi! `Next.js` definitely makes this a little tricky but I've got this working by just relying on the `defaultStore`. I am no longer using the `Provider` nor am I creating a separate `store`.

**Reply by dai-shi** [reply]
Be aware that the default store is shared across sessions and users on the server. So, _if_ you modify the store on the server, that will affect all sessions and users. If it's not changed on the server, it should be safe, but it's generally hard to follow the code and even detect it.

**Reply by anshul-kai** [reply]
Thanks for the info. I didn't even consider the possibility of this running on the server. So far, my code is strictly reserving atoms to the browser but I really appreciate the warning.

**Reply by gpxl-dev** [reply]
I'm getting the warning - I do have multiple `<Provider>`s, but none of them pass a `store` prop. 

Just to confirm:

1. If I don't explicitly pass a store to a provider, the default store is **not** used? I am confused about what is "default" about it if it isn't something you have to override?
2. Using `<Provider>` with no store is safe and will not leak data across different sessions?
3. I can essentially ignore the warning unless I start to use `getDefaultStore()` somewhere?

**Reply by dai-shi** [reply]
1: The default store is used if useAtomValue (and useAtom) hook doesn't find `<Provider>` in the component tree.
2: I believe so. Be sure to put `<Provider>` close to the root.
3: Even if you don't explicitly use `getDefaultStore()`, it can be used as it's the default. You can basically ignore the warning if it's caused by hot reload while developing.

**Reply by dbritto-dev** [reply]
@dai-shi I'm getting this issue using `<Provider>{children}</Provider>` at layout level on Nextjs App Router App

**Reply by dai-shi** [reply]
Do you see the log initially when you start the dev server? Or, it's only when HMR happens?

**Reply by dbritto-dev** [reply]
@dai-shi seems is a HMR thing, may is related to `React Refresh`

**Reply by dai-shi** [reply]
if you are editing the layout file and see the log, I think it's expected.

**Reply by dbritto-dev** [reply]
@dai-shi not layout but any page

**Reply by dai-shi** [reply]
Hm, so I guess when any page is edited, the layout is also HMR'd. Can you confirm that with adding console.log or something?

**Reply by dbritto-dev** [reply]
@dai-shi yeah, seems is also HMR'd

**Reply by dai-shi** [reply]
If `<Provider>` is HMR'd, the warning should be shown, as something unexpected can really be happen.

So the question is if there's a way to avoid HMR'ing the layout file when we only edit the page file.

**Reply by dai-shi** [reply]
Wait, that's still weird. Are you somehow using atom outside `<Provider>` tree?

**Reply by dbritto-dev** [reply]
@dai-shi no, all the components should be inside the `<Provider>` tree

**Reply by dbritto-dev** [reply]
@dai-shi I'll create a stackblitz minimal repro
